### PR TITLE
Cleanup RegionIterator to use RocksDB features directly

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

###  What have you changed?

By utilizing some recent RocksDB features, like `seek_for_prev` and iterator bound, we can simplify our `RegionIterator`. Also moved some error cases to standalone functions to reduce the hot-path code size.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test
- Integration test

### CODE REVIEW

**Comment by breeswish** [comment]
@Connor1996 I guess it is because we first have `RegionIterator`, then `seek_for_prev` and iterator upper / lower bound appears.

File: src/raftstore/store/region_snapshot.rs [hunk 1]
```diff
@@ -371,20 +313,25 @@ impl RegionIterator {
...
-                return Err(e);
-            }
+            return handle_check_key_in_region_error(e);
         }
         Ok(())
     }
 }
 
+fn handle_check_key_in_region_error(e: crate::raftstore::Error) -> Result<()> {
+    // Split out the error case to reduce hot-path code size.
```
Comment (brson) [inline src/raftstore/store/region_snapshot.rs]: If this function is here to reduce code size it should probably be `#[inline(never)]`. It's not obvious that putting the error handling in a separate function would have a performance impact, but it probably won't hurt.

File: src/storage/kv/cursor.rs [hunk 2]
```diff
@@ -346,30 +348,33 @@ impl<I: Iterator> Cursor<I> {
...
+                self.handle_error_status(e)?;
             }
             Ok(false)
         } else {
             Ok(true)
         }
     }
+
+    fn handle_error_status(&self, e: Error) -> Result<()> {
+        // Split out the error case to reduce hot-path code size.
```
Comment (brson) [inline src/storage/kv/cursor.rs]: Again seems like this function wants to be `#[inline(never)]`

**Comment by breeswish** [comment]
Follow up in https://github.com/tikv/tikv/pull/6010

**Pull request closed without merging**